### ISSUE TITLE
[bug] DeepSeek-V4-Flash hangs in PIPELINE_MODEL_PARALLEL_GROUP with TP=1/PP=2/EP=8 on 2x8 H200

### ISSUE THREAD

**Issue by WaldenLee2005** [NVIDIA-NeMo/Megatron-Bridge#5088]
### Problem

Hi Megatron team,

We are testing DeepSeek-V4-Flash with Megatron-Bridge / Megatron-Core on two 8xH200 nodes, 16 GPUs total.

The run uses:
- TP=1
- PP=2
- EP=8
- 2 nodes x 8 GPUs
- BF16 / H200
- tiny smoke dataset
- seq_len=16
- micro_batch_size=1
- grad_accum=2
- max_local_steps=1

The run gets through:
- 16-rank distributed init
- DeepSeek-V4-Flash import into Megatron-Core
- LoRA attachment
- DDP / distributed optimizer setup
- dataset loading
- first training step entry

It then hangs until the NCCL watchdog times out in the pipeline parallel group:

[PG ID 6 ... PIPELINE_MODEL_PARALLEL_GROUP Rank 1]
Watchdog caught collective operation timeout:
WorkNCCL(SeqNum=1, OpType=COALESCED, Timeout(ms)=1800000)

Control tests we ran on the same nodes:
- 16-rank NCCL all-reduce over the same two nodes passed.
- 16-rank NCCL send/recv over the same two nodes passed.
- DeepSeek-V4-Flash with EP=8 / PP=1 works on the same 16 GPUs.
- Qwen with PP=2 works on the same Yeto/Megatron path.
- The failure appears specific to DeepSeek-V4-Flash + pipeline parallelism.

Questions:
1. Is DeepSeek-V4-Flash expected to support TP=1 / PP=2 / EP=8 on 16 GPUs?
2. Or is the expected/supported DS4 shape only TP=1 / PP=4 / EP=8 on 32 GPUs?
3. If PP=2 should work, are there additional DS4-specific settings needed for pipeline parallelism, MTP, hybrid attention, or expert routing?
4. Is this likely a known Megatron-Bridge / Megatron-Core DS4 pipeline-parallel issue?

I attached the node0 and node1 logs from the failing run.

### Minimal repro

```shell
1. Create this file on both nodes:


cat > /workspace/records/smoke_chat_many.jsonl <<'EOF'
{"messages":[{"role":"user","content":"Say hello."},{"role":"assistant","content":"Hello!"}]}
{"messages":[{"role":"user","content":"What is 2+2?"},{"role":"assistant","content":"2+2 is 4."}]}
{"messages":[{"role":"user","content":"Name one color."},{"role":"assistant","content":"Blue."}]}
{"messages":[{"role":"user","content":"Write one sentence."},{"role":"assistant","content":"The test is running."}]}
{"messages":[{"role":"user","content":"Say hello again."},{"role":"assistant","content":"Hello again!"}]}
{"messages":[{"role":"user","content":"What is 3+3?"},{"role":"assistant","content":"3+3 is 6."}]}
{"messages":[{"role":"user","content":"Name another color."},{"role":"assistant","content":"Green."}]}
{"messages":[{"role":"user","content":"Write another sentence."},{"role":"assistant","content":"The smoke test is still running."}]}
EOF

2. Run this on node 1
export MASTER_ADDR=<NODE_1_TAILSCALE_IP>
export NCCL_SOCKET_IFNAME=tailscale0
export GLOO_SOCKET_IFNAME=tailscale0
export NCCL_DEBUG=WARN
export TORCH_DISTRIBUTED_DEBUG=DETAIL

cd /workspace/yeto

torchrun --nnodes=2 --node_rank=0 --nproc_per_node=8 \
  --master_addr=$MASTER_ADDR --master_port=29540 \
  -m yeto.megatron.learner \
  --model deepseek4flash \
  --trust-remote-code \
  --data /workspace/records/smoke_chat_many.jsonl \
  --syncer none \
  --island-backend megatron \
  --expert-parallel 8 \
  --tensor-parallel 1 \
  --pipeline-parallel 2 \
  --max-rows 64 \
  --seq-len 16 \
  --grad-accum 2 \
  --micro-batch-size 1 \
  --max-local-steps 1 \
  --train-on all \
  2>&1 | tee /workspace/records/deepseek4flash-2x8h200-pp2-node0-port29540.log

3. Run this on node 2:
export MASTER_ADDR=<NODE_1_TAILSCALE_IP>
export NCCL_SOCKET_IFNAME=tailscale0
export GLOO_SOCKET_IFNAME=tailscale0
export NCCL_DEBUG=WARN
export TORCH_DISTRIBUTED_DEBUG=DETAIL

cd /workspace/yeto

torchrun --nnodes=2 --node_rank=1 --nproc_per_node=8 \
  --master_addr=$MASTER_ADDR --master_port=29540 \
  -m yeto.megatron.learner \
  --model deepseek4flash \
  --trust-remote-code \
  --data /workspace/records/smoke_chat_many.jsonl \
  --syncer none \
  --island-backend megatron \
  --expert-parallel 8 \
  --tensor-parallel 1 \
  --pipeline-parallel 2 \
  --max-rows 64 \
  --seq-len 16 \
  --grad-accum 2 \
  --micro-batch-size 1 \
  --max-local-steps 1 \
  --train-on all \
  2>&1 | tee /workspace/records/deepseek4flash-2x8h200-pp2-node1-port29540.log

Result Observed:
[PG ID 6 ... PIPELINE_MODEL_PARALLEL_GROUP Rank 1]
Watchdog caught collective operation timeout:
WorkNCCL(SeqNum=1, OpType=COALESCED, Timeout(ms)=1800000)
```

### Expected behavior

DeepSeek-V4-Flash should either:

1. Complete at least one training step successfully with `TP=1/PP=2/EP=8`, writing the adapter/checkpoint artifact and exiting cleanly, or

2. Fail early with a clear validation error if `TP=1/PP=2/EP=8` is not a supported DeepSeek-V4-Flash parallelism configuration.

It should not hang for 30 minutes and then fail with an NCCL watchdog timeout in `PIPELINE_MODEL_PARALLEL_GROUP`.

### Affected area

area:training

### Regression?

Not sure

### Environment

Branch / commit:
- https://github.com/WaldenLee2005/yeto/tree/megatron-e2e-test
- Yeto branch: `megatron-e2e-test`
- Latest Yeto commit used: `5934b93 Deduplicate Megatron PP adapter export`
- Megatron-Bridge: NVIDIA Megatron-Bridge from `nvcr.io/nvidia/nemo:25.09` container, with newer source checkout used for DS4 import testing

Container or environment:
- Container: `nvcr.io/nvidia/nemo:25.09`
- Run inside Docker with `--gpus all --ipc=host --network host`
- Inter-node networking over Tailscale
- `NCCL_SOCKET_IFNAME=tailscale0`
- `GLOO_SOCKET_IFNAME=tailscale0`

GPU / machine:
- 2 nodes x 8 GPUs
- 16 total NVIDIA H200 GPUs
- Each node sees 8x H200
- Same nodes passed 16-rank NCCL all-reduce and send/recv tests

Python / CUDA / PyTorch:
- Python: 3.12
- PyTorch: `2.8.0a0+5228986c39.nv25.06`
- CUDA runtime from PyTorch: CUDA 12.x / NeMo 25.09 environment
- NCCL: `2.27.3+cuda12.9`
- Transformers: DS4 required newer Transformers support during testing; packaged container version did not recognize `deepseek_v4`

### Logs

```shell
The run hangs after entering the first training step and then times out in the pipeline parallel group.

Relevant node0 log:


2026-07-25 00:57:09,321 megatron INFO 258 adapter tensors -> 8 fragments
2026-07-25 00:57:16,783 megatron INFO Using reduce-scatter for gradient reductions because self.ddp_config.use_distributed_optimizer=True
2026-07-25 00:57:17,603 megatron INFO HTTP Request: HEAD https://s3.amazonaws.com/datasets.huggingface.co/datasets/datasets/json/json.py "HTTP/1.1 200 OK"
Generating train split: 64 examples [00:00, 16694.79 examples/s]
[rank0]:[E725 01:27:17.961331563 ProcessGroupNCCL.cpp:1772] [PG ID 0 PG GUID 0(default_pg) Rank 0] Received a dump signal due to a collective timeout from rank 12. Last enqueued NCCL work: 8, last completed NCCL work: 8. This is most likely caused by incorrect usages of collectives, e.g. wrong sizes used across ranks, the order of collectives is not same for all ranks or the scheduled collective, for some reason, didn't run.

node 1 log: 
[rank12]:[E725 01:28:17.462229054 ProcessGroupNCCL.cpp:1923] [PG ID 6 PG GUID 61(PIPELINE_MODEL_PARALLEL_GROUP) Rank 1] Process group watchdog thread terminated with exception: [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=1, OpType=COALESCED, NumelIn=18446744073709551615, NumelOut=18446744073709551615, Timeout(ms)=1800000) ran for 1800046 milliseconds before timing out.
[rank12]:[E725 01:28:17.460994934 ProcessGroupNCCL.cpp:707] [Rank 1] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[rank12]:[E725 01:28:17.461014442 ProcessGroupNCCL.cpp:721] [Rank 1] To avoid data inconsistency, we are taking the entire process down.
```

**Comment by yaoyu-33** [comment]
@weijiac0619  share PP run log if successful.

**Comment by weijiac0619** [comment]
Tested PP=2 EP=8 on CW-DFW H100 cluster and there is no hang.
                                                                                                                                                                                                                                                 
  Settings:                                                                       
  - Megatron-Bridge: 5151999ad (main, 2026-07-27)
  - Megatron-Core: 9304a2564 (dev, 2026-07-27)
  - Config: TP=1, PP=2, EP=8, 16× H100 GPUs, 2 nodes

  Result: 5/5 iterations completed cleanly, no PIPELINE_MODEL_PARALLEL_GROUP hang:

  [2026-07-27 19:08:04] iteration  1/5 | elapsed (ms): 63420 | lm loss: 5.552440 | grad norm: 0.725
  [2026-07-27 19:08:05] iteration  2/5 | elapsed (ms):  95.6 | lm loss: 5.552458 | grad norm: 0.726
  [2026-07-27 19:08:05] iteration  3/5 | elapsed (ms):  43.8 | lm loss: 5.519175 | grad norm: 0.725
  [2026-07-27 19:08:05] iteration  4/5 | elapsed (ms):  34.8 | lm loss: 5.485632 | grad norm: 0.725
  [2026-07-27 19:08:05] iteration  5/5 | elapsed (ms):  35.1 | lm loss: 5.466413 | grad norm: 0.723
  [smoke] PASSED: 5 steps completed.

  PP layout for 2 layers / PP=2: [['embedding', 'decoder'], ['decoder', 'loss']]

**Comment by meatybobby** [comment]
@WaldenLee2005 Could you try the same commit and config which @weijiac0619 used?

**Comment by WaldenLee2005** [comment]
> [@WaldenLee2005](https://github.com/WaldenLee2005) Could you try the same commit and config which [@weijiac0619](https://github.com/weijiac0619) used?

Yeah, got it working, turns out it was an issue on our end. Thank you for the confirmation that your end worked!
